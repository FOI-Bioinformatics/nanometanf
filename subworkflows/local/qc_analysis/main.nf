/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    SUBWORKFLOW: QC_ANALYSIS
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Multi-tool quality control analysis with nanopore optimization

    Supported QC tools:
    - chopper: Nanopore-native Rust-based filtering (7x faster than NanoFilt)
    - fastp: General-purpose QC with rich reporting
    - filtlong: Nanopore-optimized length-weighted quality filtering

    Future QC tools (ready to implement):
    - nanoq: Nanopore quality assessment

    Features:
    - Tool-agnostic interface
    - Nanopore-specific optimizations
    - Consistent output standardization
    - Easy addition of new QC tools
----------------------------------------------------------------------------------------
*/

include { FASTP                   } from "${projectDir}/modules/nf-core/fastp/main"
include { FILTLONG                } from "${projectDir}/modules/nf-core/filtlong/main"
include { CHOPPER                 } from "${projectDir}/modules/nf-core/chopper/main"
include { PORECHOP_PORECHOP       } from "${projectDir}/modules/nf-core/porechop/porechop/main"
include { NANOPLOT                } from "${projectDir}/modules/nf-core/nanoplot/main"
include { FASTQC                  } from "${projectDir}/modules/nf-core/fastqc/main"
include { SEQKIT_STATS            } from "${projectDir}/modules/nf-core/seqkit/stats/main"
include { SEQKIT_MERGE_STATS      } from "${projectDir}/modules/local/seqkit_merge_stats/main"

workflow QC_ANALYSIS {

    take:
    ch_reads     // channel: [ val(meta), path(reads) ]

    main:
    ch_versions = Channel.empty()
    ch_qc_reads = Channel.empty()
    ch_qc_reports = Channel.empty()
    ch_qc_logs = Channel.empty()
    ch_qc_json = Channel.empty()
    ch_fastqc_html = Channel.empty()
    ch_seqkit_stats = Channel.empty()
    
    // Set QC tool and validate parameters
    def qc_tool = params.qc_tool ?: 'fastp'
    def enable_adapter_trimming = params.enable_adapter_trimming ?: false
    
    //
    // OPTIONAL: Adapter trimming with PORECHOP (nanopore-specific)
    //
    if (enable_adapter_trimming && qc_tool == 'filtlong') {
        PORECHOP_PORECHOP (
            ch_reads
        )
        ch_versions = ch_versions.mix(PORECHOP_PORECHOP.out.versions)
        ch_adapter_trimmed = PORECHOP_PORECHOP.out.reads
    } else {
        ch_adapter_trimmed = ch_reads
    }
    
    //
    // BRANCH: Route to appropriate QC tool
    //
    switch(qc_tool) {
        case 'fastp':
            //
            // MODULE: Run FASTP for general-purpose quality filtering and QC
            //
            FASTP (
                ch_adapter_trimmed,
                [],           // adapter_fasta
                false,        // discard_trimmed_pass  
                false,        // save_trimmed_fail
                false         // save_merged
            )
            ch_versions = ch_versions.mix(FASTP.out.versions)
            
            // Collect standardized outputs
            ch_qc_reads = FASTP.out.reads
            ch_qc_reports = FASTP.out.html
            ch_qc_logs = FASTP.out.log
            ch_qc_json = FASTP.out.json
            break
            
        case 'filtlong':
            //
            // MODULE: Run FILTLONG for nanopore-optimized quality filtering
            //
            // FILTLONG expects shortreads and longreads, for nanopore-only we pass empty for shortreads
            ch_filtlong_input = ch_adapter_trimmed.map { meta, reads ->
                [meta, [], reads]  // [meta, shortreads=empty, longreads=reads]
            }
            
            FILTLONG (
                ch_filtlong_input
            )
            ch_versions = ch_versions.mix(FILTLONG.out.versions)
            
            //
            // Enhanced reporting for FILTLONG: Add FastQC and SeqKit stats
            //
            
            // MODULE: Run FastQC on filtered reads for comprehensive HTML reporting
            FASTQC (
                FILTLONG.out.reads
            )
            ch_versions = ch_versions.mix(FASTQC.out.versions)
            ch_fastqc_html = FASTQC.out.html
            
            // MODULE: Run SeqKit stats for detailed sequence statistics  
            SEQKIT_STATS (
                FILTLONG.out.reads
            )
            ch_versions = ch_versions.mix(SEQKIT_STATS.out.versions)
            ch_seqkit_stats = SEQKIT_STATS.out.stats
            
            // Collect standardized outputs
            ch_qc_reads = FILTLONG.out.reads
            ch_qc_reports = ch_fastqc_html              // Use FastQC HTML reports for FILTLONG
            ch_qc_logs = FILTLONG.out.log
            ch_qc_json = ch_seqkit_stats                // Use SeqKit stats as JSON-like structured output
            break

        case 'chopper':
            //
            // MODULE: Run CHOPPER for nanopore-native quality filtering
            //
            // CHOPPER is Rust-based, 7x faster than NanoFilt, optimized for nanopore data
            CHOPPER (
                ch_adapter_trimmed,
                []  // No contamination filtering fasta
            )
            ch_versions = ch_versions.mix(CHOPPER.out.versions)

            //
            // Enhanced reporting for CHOPPER: Add FastQC and SeqKit stats
            //

            // MODULE: Run FastQC on filtered reads for comprehensive HTML reporting
            FASTQC (
                CHOPPER.out.fastq
            )
            ch_versions = ch_versions.mix(FASTQC.out.versions)
            ch_fastqc_html = FASTQC.out.html

            // MODULE: Run SeqKit stats for detailed sequence statistics
            SEQKIT_STATS (
                CHOPPER.out.fastq
            )
            ch_versions = ch_versions.mix(SEQKIT_STATS.out.versions)
            ch_seqkit_stats = SEQKIT_STATS.out.stats

            // Collect standardized outputs
            ch_qc_reads = CHOPPER.out.fastq
            ch_qc_reports = ch_fastqc_html              // Use FastQC HTML reports for CHOPPER
            ch_qc_logs = Channel.empty()                // CHOPPER has no log output
            ch_qc_json = ch_seqkit_stats                // Use SeqKit stats as JSON-like structured output
            break

        // Future QC tools to be added here:
        // case 'nanoq':
        //     NANOQ(ch_adapter_trimmed)
        //     break

        default:
            error "Unsupported QC tool: ${qc_tool}. Currently supported: fastp, filtlong, chopper"
    }

    //
    // OPTIONAL: Incremental QC statistics aggregation (PromethION optimization)
    //
    // For tools that use SEQKIT_STATS (chopper, filtlong), aggregate batch-level stats
    // into cumulative statistics when incremental mode is enabled
    //
    def enable_incremental = params.qc_enable_incremental ?: false
    def ch_final_seqkit_stats = ch_seqkit_stats

    if (enable_incremental && (qc_tool == 'chopper' || qc_tool == 'filtlong')) {
        log.info "Using incremental QC statistics aggregation for ${qc_tool}"

        // Group batch-level seqkit stats by sample ID
        def ch_grouped_batch_stats = ch_seqkit_stats.groupTuple(by: 0)

        // Merge batch statistics into cumulative statistics
        SEQKIT_MERGE_STATS(
            ch_grouped_batch_stats
        )
        ch_versions = ch_versions.mix(SEQKIT_MERGE_STATS.out.versions)

        // Use cumulative stats instead of batch stats
        ch_final_seqkit_stats = SEQKIT_MERGE_STATS.out.cumulative_stats
    }

    //
    // MODULE: Run NanoPlot for nanopore-specific QC visualization (conditional in real-time mode)
    //
    // NanoPlot optimization for real-time processing:
    // - Skip intermediate batches if nanoplot_realtime_skip_intermediate = true
    // - Run every N batches if nanoplot_batch_interval is set
    // - Always run on final batch (when is_final_batch = true in meta)
    //
    def skip_nanoplot = params.skip_nanoplot ?: false
    def is_realtime = params.realtime_mode ?: false
    def skip_intermediate = params.nanoplot_realtime_skip_intermediate ?: true
    def batch_interval = params.nanoplot_batch_interval ?: 10

    def ch_nanoplot_input = ch_qc_reads
    def ch_nanoplot_html = Channel.empty()
    def ch_nanoplot_txt = Channel.empty()
    def ch_nanoplot_png = Channel.empty()

    if (!skip_nanoplot) {
        // Apply real-time optimizations
        if (is_realtime && skip_intermediate) {
            log.info "Real-time mode: NanoPlot will run every ${batch_interval} batches and on final batch"

            // Filter channel to only include samples that should run NanoPlot
            ch_nanoplot_input = ch_qc_reads.filter { meta, reads ->
                // Always run on final batch
                if (meta.is_final_batch == true) {
                    log.info "Running NanoPlot for ${meta.id} (final batch)"
                    return true
                }

                // Run every N batches based on batch_id
                if (meta.batch_id != null) {
                    def batch_num = meta.batch_id instanceof String ?
                        Integer.parseInt(meta.batch_id.replaceAll(/\D/, '')) : meta.batch_id

                    if (batch_num % batch_interval == 0) {
                        log.info "Running NanoPlot for ${meta.id} (batch ${batch_num} - interval milestone)"
                        return true
                    }
                }

                // Skip this batch
                log.debug "Skipping NanoPlot for ${meta.id} (intermediate batch)"
                return false
            }
        }

        // Run NanoPlot on filtered samples
        NANOPLOT (
            ch_nanoplot_input
        )
        ch_versions = ch_versions.mix(NANOPLOT.out.versions)
        ch_nanoplot_html = NANOPLOT.out.html
        ch_nanoplot_txt = NANOPLOT.out.txt
        ch_nanoplot_png = NANOPLOT.out.png
    } else {
        log.info "NanoPlot is disabled (skip_nanoplot = true)"
    }

    emit:
    reads        = ch_qc_reads            // channel: [ val(meta), path(reads) ] - QC'd reads
    qc_reports   = ch_qc_reports          // channel: [ val(meta), path(html) ] - QC HTML reports (tool-specific)
    qc_logs      = ch_qc_logs             // channel: [ val(meta), path(log) ] - QC log files
    qc_json      = ch_qc_json             // channel: [ val(meta), path(json) ] - QC JSON reports (if available)
    nanoplot     = ch_nanoplot_html       // channel: [ val(meta), path(html) ] - NanoPlot visualization (conditional in real-time mode)
    nanoplot_txt = ch_nanoplot_txt        // channel: [ val(meta), path(txt) ] - NanoPlot summary stats (conditional, for MultiQC)
    nanoplot_png = ch_nanoplot_png        // channel: [ val(meta), path(png) ] - NanoPlot plots (conditional)
    qc_tool_used = Channel.value(qc_tool) // channel: val(qc_tool_name) - Tool identification
    versions     = ch_versions            // channel: [ path(versions.yml) ]

    // Enhanced reporting outputs
    fastqc_html  = ch_fastqc_html         // channel: [ val(meta), path(html) ] - FastQC HTML reports (FILTLONG/CHOPPER enhancement)
    seqkit_stats = ch_final_seqkit_stats  // channel: [ val(meta), path(txt) ] - SeqKit detailed statistics (cumulative if incremental mode)

    // Legacy outputs for backward compatibility
    fastp_json   = qc_tool == 'fastp' ? ch_qc_json : Channel.empty()     // channel: [ val(meta), path(json) ]
    fastp_html   = qc_tool == 'fastp' ? ch_qc_reports : Channel.empty()  // channel: [ val(meta), path(html) ]
    fastp_log    = qc_tool == 'fastp' ? ch_qc_logs : Channel.empty()     // channel: [ val(meta), path(log) ]
}