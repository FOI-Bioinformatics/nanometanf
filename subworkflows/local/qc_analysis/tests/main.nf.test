nextflow_workflow {

    name "Test QC_ANALYSIS"
    script "../main.nf"
    workflow "QC_ANALYSIS"

    test("Should perform basic FASTP quality control") {

        setup {
            """
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/basic_sample.fastq.gz
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'basic_qc_test', single_end: true],
                        file('$outputDir/basic_sample.fastq.gz')
                    ]
                ]
                """
            }
            
            params {
                qc_tool = 'fastp'
                skip_fastp = false
                skip_nanoplot = true
                skip_fastqc = true
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '2.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.qc_reads
            assert workflow.out.qc_reports
            
            // Verify FASTP processing
            def qc_reads = workflow.out.qc_reads.toList()
            assert qc_reads.size() >= 1
            
            def qc_reports = workflow.out.qc_reports.toList()
            assert qc_reports.size() >= 1
        }
    }

    test("Should perform comprehensive nanopore QC with multiple tools") {

        setup {
            """
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/comprehensive_sample.fastq.gz
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'comprehensive_qc_test', single_end: true],
                        file('$outputDir/comprehensive_sample.fastq.gz')
                    ]
                ]
                """
            }
            
            params {
                qc_tool = 'fastp'
                skip_fastp = false
                skip_nanoplot = false
                skip_fastqc = false
                enable_filtlong = true
                enable_porechop = true
                enable_seqkit_stats = true
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '3.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.qc_reads
            assert workflow.out.qc_reports
            assert workflow.out.nanoplot_reports
            
            // Verify multiple QC tools ran
            def qc_reads = workflow.out.qc_reads.toList()
            assert qc_reads.size() >= 1
            
            def qc_reports = workflow.out.qc_reports.toList()
            assert qc_reports.size() >= 1
            
            // Verify nanopore-specific QC
            if (workflow.out.nanoplot_reports) {
                def nanoplot_reports = workflow.out.nanoplot_reports.toList()
                assert nanoplot_reports.size() >= 1
            }
        }
    }

    test("Should handle FILTLONG nanopore-optimized filtering") {

        setup {
            """
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/filtlong_sample.fastq.gz
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'filtlong_test', single_end: true],
                        file('$outputDir/filtlong_sample.fastq.gz')
                    ]
                ]
                """
            }
            
            params {
                qc_tool = 'filtlong'
                filtlong_min_length = 1000
                filtlong_keep_percent = 90
                filtlong_target_bases = 500000000
                skip_fastp = true
                skip_nanoplot = true
                skip_fastqc = true
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '2.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.qc_reads
            
            // Verify FILTLONG processing
            def qc_reads = workflow.out.qc_reads.toList()
            assert qc_reads.size() >= 1
        }
    }

    test("Should perform adapter trimming with PORECHOP") {

        setup {
            """
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/adapter_sample.fastq.gz
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'porechop_test', single_end: true],
                        file('$outputDir/adapter_sample.fastq.gz')
                    ]
                ]
                """
            }
            
            params {
                qc_tool = 'fastp'
                enable_porechop = true
                porechop_discard_middle = true
                porechop_extra_end_trim = 2
                skip_nanoplot = true
                skip_fastqc = true
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '2.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.qc_reads
            
            // Verify adapter trimming occurred
            def qc_reads = workflow.out.qc_reads.toList()
            assert qc_reads.size() >= 1
        }
    }

    test("Should generate comprehensive nanopore statistics") {

        setup {
            """
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/stats_sample.fastq.gz
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'stats_test', single_end: true],
                        file('$outputDir/stats_sample.fastq.gz')
                    ]
                ]
                """
            }
            
            params {
                qc_tool = 'fastp'
                enable_seqkit_stats = true
                skip_nanoplot = false
                nanoplot_format = 'png'
                skip_fastqc = true
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '2.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.qc_reads
            assert workflow.out.qc_reports
            assert workflow.out.stats_reports
            
            // Verify statistics generation
            def stats_reports = workflow.out.stats_reports.toList()
            assert stats_reports.size() >= 1
            
            // Verify NanoPlot visualization
            if (workflow.out.nanoplot_reports) {
                def nanoplot_reports = workflow.out.nanoplot_reports.toList()
                assert nanoplot_reports.size() >= 1
            }
        }
    }

    test("Should handle multiple samples in batch processing") {

        setup {
            """
            # Create multiple test samples
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/batch_sample1.fastq.gz
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/batch_sample2.fastq.gz
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/batch_sample3.fastq.gz
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'batch_sample1', single_end: true],
                        file('$outputDir/batch_sample1.fastq.gz')
                    ],
                    [
                        [id: 'batch_sample2', single_end: true],
                        file('$outputDir/batch_sample2.fastq.gz')
                    ],
                    [
                        [id: 'batch_sample3', single_end: true],
                        file('$outputDir/batch_sample3.fastq.gz')
                    ]
                ]
                """
            }
            
            params {
                qc_tool = 'fastp'
                skip_nanoplot = true
                skip_fastqc = true
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '3.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.qc_reads
            assert workflow.out.qc_reports
            
            // Verify batch processing
            def qc_reads = workflow.out.qc_reads.toList()
            assert qc_reads.size() >= 3
            
            def qc_reports = workflow.out.qc_reports.toList()
            assert qc_reports.size() >= 3
        }
    }

    test("Should handle quality-based filtering with custom thresholds") {

        setup {
            """
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/quality_sample.fastq.gz
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'quality_filter_test', single_end: true],
                        file('$outputDir/quality_sample.fastq.gz')
                    ]
                ]
                """
            }
            
            params {
                qc_tool = 'fastp'
                fastp_qualified_quality_phred = 15
                fastp_unqualified_percent_limit = 20
                fastp_length_required = 1000
                fastp_cut_front_window_size = 4
                fastp_cut_front_mean_quality = 20
                skip_nanoplot = true
                skip_fastqc = true
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '2.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.qc_reads
            assert workflow.out.qc_reports
            
            // Verify quality filtering
            def qc_reads = workflow.out.qc_reads.toList()
            assert qc_reads.size() >= 1
            
            def qc_reports = workflow.out.qc_reports.toList()
            assert qc_reports.size() >= 1
        }
    }

    test("Should handle FastQC analysis for general quality assessment") {

        setup {
            """
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/fastqc_sample.fastq.gz
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'fastqc_test', single_end: true],
                        file('$outputDir/fastqc_sample.fastq.gz')
                    ]
                ]
                """
            }
            
            params {
                qc_tool = 'fastp'
                skip_fastp = true
                skip_nanoplot = true
                skip_fastqc = false
                fastqc_adapters = null
                fastqc_limits = null
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '2.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.fastqc_reports
            
            // Verify FastQC analysis
            def fastqc_reports = workflow.out.fastqc_reports.toList()
            assert fastqc_reports.size() >= 1
        }
    }

    test("Should handle edge cases with small or malformed files") {

        setup {
            """
            # Create problematic files
            echo -e "@read1\nATCG\n+\n!!!!" | gzip > $outputDir/tiny_sample.fastq.gz
            touch $outputDir/empty_sample.fastq.gz
            echo "not a fastq file" > $outputDir/malformed.txt
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'tiny_test', single_end: true],
                        file('$outputDir/tiny_sample.fastq.gz')
                    ],
                    [
                        [id: 'empty_test', single_end: true],
                        file('$outputDir/empty_sample.fastq.gz')
                    ]
                ]
                """
            }
            
            params {
                qc_tool = 'fastp'
                skip_nanoplot = true
                skip_fastqc = true
                fastp_dont_eval_duplication = true
                max_cpus = 1
                max_memory = '2.GB'
                max_time = '1.min'
            }
        }

        then {
            // Should handle edge cases gracefully
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            
            if (workflow.success) {
                // May have limited outputs due to edge cases
                if (workflow.out.qc_reads) {
                    def qc_reads = workflow.out.qc_reads.toList()
                    assert qc_reads.size() >= 0
                }
            }
        }
    }

    test("Should run minimal QC with skip options") {

        setup {
            """
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/minimal_sample.fastq.gz
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'minimal_qc_test', single_end: true],
                        file('$outputDir/minimal_sample.fastq.gz')
                    ]
                ]
                """
            }
            
            params {
                qc_tool = 'fastp'
                skip_fastp = false
                skip_nanoplot = true
                skip_fastqc = true
                enable_filtlong = false
                enable_porechop = false
                enable_seqkit_stats = false
                fastp_save_trimmed_fail = false
                fastp_save_merged = false
                max_cpus = 1
                max_memory = '2.GB'
                max_time = '1.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.qc_reads
            
            // Verify minimal processing
            def qc_reads = workflow.out.qc_reads.toList()
            assert qc_reads.size() >= 1
        }
    }

    test("Should validate QC workflow with no input files") {

        when {
            workflow {
                """
                input[0] = []
                """
            }
            
            params {
                qc_tool = 'fastp'
                skip_fastp = false
                max_cpus = 1
                max_memory = '1.GB'
                max_time = '30.s'
            }
        }

        then {
            // Should handle empty input gracefully
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            
            if (workflow.success) {
                // Empty outputs expected
                if (workflow.out.qc_reads) {
                    def qc_reads = workflow.out.qc_reads.toList()
                    assert qc_reads.size() == 0
                }
            }
        }
    }
}