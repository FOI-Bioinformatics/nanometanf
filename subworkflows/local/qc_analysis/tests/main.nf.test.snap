{
    "Should handle FastQC analysis for general quality assessment": {
        "content": [
            [
                "versions.yml:md5,289c30c0c8f1d0b00532329504572f16",
                "versions.yml:md5,b20891e5d879f88cd6f63e9573d45c31"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-14T20:29:16.585064"
    },
    "Should handle quality-based filtering with custom thresholds": {
        "content": [
            [
                "versions.yml:md5,289c30c0c8f1d0b00532329504572f16",
                "versions.yml:md5,b20891e5d879f88cd6f63e9573d45c31"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-14T20:29:07.346736"
    },
    "Should perform comprehensive nanopore QC with multiple tools": {
        "content": [
            [
                "versions.yml:md5,289c30c0c8f1d0b00532329504572f16",
                "versions.yml:md5,b20891e5d879f88cd6f63e9573d45c31"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-14T20:28:35.962374"
    },
    "Should perform basic FASTP quality control": {
        "content": [
            [
                "versions.yml:md5,289c30c0c8f1d0b00532329504572f16",
                "versions.yml:md5,b20891e5d879f88cd6f63e9573d45c31"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-14T20:19:56.74021"
    },
    "Should run minimal QC with skip options": {
        "content": [
            [
                "versions.yml:md5,289c30c0c8f1d0b00532329504572f16",
                "versions.yml:md5,b20891e5d879f88cd6f63e9573d45c31"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-14T20:29:27.84905"
    },
    "Should generate comprehensive nanopore statistics": {
        "content": [
            [
                "versions.yml:md5,278de80cb05a3c6ded00c0b6ccd67373",
                "versions.yml:md5,289c30c0c8f1d0b00532329504572f16",
                "versions.yml:md5,f032c923a92969ab13137db3483fb44f",
                "versions.yml:md5,fc9c37561b82c26a4f7126eff84900de"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-15T17:02:45.547267"
    },
    "Should perform adapter trimming with PORECHOP": {
        "content": [
            [
                "versions.yml:md5,289c30c0c8f1d0b00532329504572f16",
                "versions.yml:md5,b20891e5d879f88cd6f63e9573d45c31"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-14T20:28:47.654786"
    }
}